### DISCUSSION TITLE
Should teal use Cyan instead of make?

### DISCUSSION THREAD

**Discussion by lenscas** [teal-language/tl#600] [General]
With Cyan supporting to run scripts at various times during the build process, it sounds like it is possible for tl to use cyan instead of make?

I think that using Cyan would be an improvement as it means that `tl` itself would mimic more closely how teal projects would be liked to setup. It also means that the `tl` project would benefit from any improvements to `cyan` (for example if `cyan` gets support for getting the correct type definitions files for your dependencies then `tl.tl` has a more clear path of adding dependencies than it has right now.)

It also means that changes to the build script don't require `make` knowledge but instead requires you to write normal teal/lua code instead.

**Comment by hishamhm** [comment]
That's a super fair question and dogfooding is a super important step in making developer tooling better... but I'm a bit afraid of the chicken-and-egg problem for my _own_ dev experience. It's already loopy enough that I'm developing the compiler using the compiler and I often break it and end up editing the generated tl.lua to unbreak it and keep coding... if that ended up breaking the build tool as well that would put me in an even messier place :laughing: 

For that bootstrapping reason I tend to prefer using a plain Makefile for tl itself (also, the Makefile does more than just build it; it builds it then tries to build it again using the built version, and rolls back if the second-level build fails — that's a super-specific-compiler-bootstrap-need which I don't think belongs in Cyan anyway).

**Comment by lenscas** [comment]
I think that cyan is able to create that same behaviour using pre compile
and post compile scripts. Do a backup in a pre compile script and the
busted and roll back during a post compile script.

Cyan is already compiled to lua so not sure how teal breaking can break
cyan unless you are working on the api used by cyan. But even then using
cyan is a benefit as it means you have to keep cyan updated with any
changes in the api it uses.

Op di 20 dec. 2022 19:47 schreef Hisham Muhammad ***@***.***>:
